consensus/consensus_v2.go

```diff
@@ -205,6 +211,102 @@ func (consensus *Consensus) onAnnounce(msg *msg_pb.Message) {
 				Msg("[OnAnnounce] Block content is not verified successfully")
 			return
 		}
+
+		if consensus.ShardID == 0 {
+			//TODO: HB fix this
+			if recvMsg.BlockNum%core.ShardingSchedule.BlocksPerEpoch() == 0 {
+				consensus.generateNewVrf = true
+				consensus.generateNewVdf = true
+
+				consensus.vrfBlockNumbers = nil
+				consensus.ChainReader.WriteEpochVrfBlockNums(headerObj.Epoch, consensus.vrfBlockNumbers)
+			}
+
+			zeroVrfBytes := make([]byte, 32)
+			if !bytes.Equal(headerObj.Vrf[:], zeroVrfBytes) {
+				vrfPk := vrf_bls.NewVRFVerifier(consensus.LeaderPubKey)
+
+				var blockHash [32]byte
+				previousHeader := consensus.ChainReader.GetHeaderByNumber(headerObj.Number.Uint64() - 1)
+				previousHash := previousHeader.Hash()
+				copy(blockHash[:], previousHash[:])
+
+				hash, err := vrfPk.ProofToHash(blockHash[:], headerObj.VrfProof[:])
+				if err != nil {
+					consensus.getLogger().Warn().
+						Err(err).
+						Str("MsgBlockNum", headerObj.Number.String()).
+						Msg("[OnAnnounce] VRF verification error")
+					return
+				}
+
+				if hash != headerObj.Vrf {
+					consensus.getLogger().Warn().
+						Str("MsgBlockNum", headerObj.Number.String()).
+						Msg("[OnAnnounce] VRF proof is not valid")
+					return
+				}
+
+				consensus.vrfBlockNumbers, _ = consensus.ChainReader.ReadEpochVrfBlockNums(headerObj.Epoch)
```

Comment:
If we always read from disk. We can make "consensus.vrfBlockNumbers" a local variable.

Reply:
My initial implementation is making it a local variable but the values will lost in case  the node can is killed and later restarted.   So I used local cache/db for these values.  The values also need to be updated during node sync.

Reply:
use local variables and db/cache for them